test/generators/gen_rules_lib.ml

```diff
@@ -10,154 +10,98 @@ type test_case = {
-  List
-    ([
-       Atom "rule";
-       List [ Atom "target"; Atom (Fpath.basename target_path) ];
-       List [ Atom "deps"; Atom (Fpath.basename dep_path) ];
-       List
-         [
-           Atom "action";
-           List
-             [
-               Atom "run";
-               Atom "odoc";
-               Atom "compile";
-               Atom "-o";
-               Atom "%{target}";
-               Atom "%{deps}";
-             ];
-         ];
-     ]
-    @ enabledif)
-
-let odocl_target_rule enabledif dep_path target_path =
-  List
-    ([
-       Atom "rule";
-       List [ Atom "target"; Atom (Fpath.basename target_path) ];
-       List [ Atom "deps"; Atom (Fpath.basename dep_path) ];
-       List
-         [
-           Atom "action";
-           List
-             [
-               Atom "run";
-               Atom "odoc";
-               Atom "link";
-               Atom "-o";
-               Atom "%{target}";
-               Atom "%{deps}";
-             ];
-         ];
-     ]
-    @ enabledif)
-
-let mld_odoc_target_rule enabledif dep_path target_path =
-  List
-    ([
-       Atom "rule";
-       List [ Atom "target"; Atom (Fpath.basename target_path) ];
-       List [ Atom "deps"; Atom (Fpath.to_string dep_path) ];
-       List
-         [
-           Atom "action";
-           List
-             [
-               Atom "run";
-               Atom "odoc";
-               Atom "compile";
-               Atom "-o";
-               Atom "%{target}";
-               Atom "%{deps}";
-             ];
-         ];
-     ]
-    @ enabledif)
-
-let file_rule { input; odoc; odocl; enabledif; _ } cmt =
-  let enabledif = render_enabledif enabledif in
-  [
-    cu_target_rule enabledif input cmt;
-    odoc_target_rule enabledif cmt odoc;
-    odocl_target_rule enabledif odoc odocl;
-  ]
-
-let mld_file_rule { input; odoc; odocl; enabledif; _ } =
-  let enabledif = render_enabledif enabledif in
-  [
-    mld_odoc_target_rule enabledif input odoc;
-    odocl_target_rule enabledif odoc odocl;
-  ]
-
-let path' () f = Filename.quote (Fpath.to_string f)
-
-let ext' () f = Filename.quote (Fpath.get_ext f)
-
-let gen_rule_for_source_file case =
-  match case.cmt with
-  | Some cmt -> file_rule case cmt
-  | None -> mld_file_rule case
 
-let odocls backend p =
-  let path = Fpath.relativize ~root:backend p in
-  match path with Some p -> p | None -> assert false
+  let subdir dir rules = List (Atom "subdir" :: Atom (arg_fpath dir) :: rules)
+end
+
+let cu_target_rule enabledif dep target =
+  Dune.simple_rule ?enabledif target
+    [ "ocamlc"; "-c"; "-bin-annot"; "-o"; "%{target}"; Dune.arg_dep dep ]
+
+let odoc_target_rule enabledif dep target =
+  Dune.simple_rule ?enabledif target
+    [ "odoc"; "compile"; "-o"; "%{target}"; Dune.arg_dep dep ]
+
+let odocl_target_rule enabledif dep target =
+  Dune.simple_rule ?enabledif target
+    [ "odoc"; "link"; "-o"; "%{target}"; Dune.arg_dep dep ]
```

Comment:
``` suggestion
let target_rule enabledif dep target cmds =
  Dune.simple_rule ?enabledif target (cmds @ [ Dune.arg_dep dep ])

let cu_target_rule enabledif dep target =
  target_rule enabledif dep target
    [ "ocamlc"; "-c"; "-bin-annot"; "-o"; "%{target}" ]

let odoc_target_rule enabledif dep target =
  target_rule enabledif dep target [ "odoc"; "compile"; "-o"; "%{target}" ]

let odocl_target_rule enabledif dep target =
  target_rule enabledif dep target [ "odoc"; "link"; "-o"; "%{target}” ]
```

Reply:
I think this is a bit too far. I like having the commands written out entirely and I'd prefer to avoid manipulating them without an abstraction on top (eg. similar to the library bos).

Reply:
sounds reasonable!